runtime/doc/lua-plugin.txt

```diff
@@ -0,0 +1,433 @@
+https://en.wikipedia.org/wiki/Domain-specific_language via a `setup` function.
+
+However, doing so means that
+
+- You will have to implement and document it yourself.
+- Users will likely face inconsistencies if another plugin has a slightly
+  different DSL.
+- |init.lua| scripts that call such a `setup` function may throw an error if
+  the plugin is not installed or disabled.
+
+As an alternative, you can provide |<Plug>| mappings to allow users to define
+their own keymaps with |vim.keymap.set()|.
+
+- This requires one line of code in user configs.
+- Even if your plugin is not installed or disabled, creating the keymap won't
+  throw an error.
+
+Another option is to simply expose a Lua function or |user-commands|.
+
+However, some benefits of |<Plug>| mappings over this are that you can
+
+- Enforce options like `expr = true`.
+- Expose functionality only for specific |map-modes|.
+- Expose different behavior for different |map-modes| with a single |<Plug>|
+  mapping, without adding impurity or complexity to the underlying Lua
+  implementation.
+
+NOTE: If you have a function that takes a large options table, creating lots
+      of |<Plug>| mappings to expose all of its uses could become
+      overwhelming. It may still be beneficial to create some for the most
+      common ones.
+
+------------------------------------------------------------------------------
+Example					      *lua-plugin-plug-mapping-example*
+
+In your plugin:
+>lua
+    vim.keymap.set("n", "<Plug>(SayHello)", function() 
+        print("Hello from normal mode") 
+    end, { noremap = true })
+
+    vim.keymap.set("v", "<Plug>(SayHello)", function() 
+        print("Hello from visual mode") 
+    end, { noremap = true })
+<
+In the user's config:
+>lua
+    vim.keymap.set({"n", "v"}, "<leader>h", "<Plug>(SayHello)")
+<
+==============================================================================
+Initialization					    *lua-plugin-initialization*
+
+Newcomers to Lua plugin development will often put all initialization logic in
+a single `setup` function, which takes a table of options.
+If you do this, users will be forced to call this function in order to use
+your plugin, even if they are happy with the default configuration.
+
+Cleanly separated configuration and smart initialization allow your plugin to 
+work out of the box.
+
+NOTE: A well designed plugin has minimal impact on startup time.
+      See also |lua-plugin-lazy-loading|.
+
+Common approaches to a cleanly separated configuration are:
+
+- A Lua function, e.g. `setup(opts)` or `configure(opts)`, which only overrides the
+  default configuration and does not contain any initialization logic.
+- A Vimscript compatible table (e.g. in the |vim.g| or |vim.b| namespace) that your
+  plugin reads from and validates at initialization time.
+  See also |lua-vim-variables|.
+
+Typically, automatic initialization logic is done in a |plugin| or |ftplugin|
+script. See also |'runtimepath'|.
```

Comment:
Although technically correct (newcomers indeed often do that), I would like to ask to rephrase the "intro" or remove it completely due to implied negative connotations ("newcomers often don't know what they are doing and they do this, so ...").

Same with using "Cleanly separated ..." (implying that "not separated is not clean"). I'd suggest using either "explicitly separated", "intentionally separated", or just "separated" without any description.

Here is my suggestion:

```suggestion
Separated configuration and smart initialization allow your plugin to work out
of the box. Common approaches are:

- A Lua function, e.g. `setup(opts)` or `configure(opts)`, which only overrides
  the default configuration and does not contain any initialization logic.
- A Vimscript compatible table (e.g. in the |vim.g| or |vim.b| namespace) that
  your plugin reads from and validates at initialization time.
  See also |lua-vim-variables|.

Typically, automatic initialization logic is done in a |plugin| or |ftplugin|
script. See also |'runtimepath'|.

If you want to force users to explicitly enable plugin functionality, using
`setup(opts)` function for both initialization and configuration is a common
practice.

NOTE: A well designed plugin has minimal impact on startup time.
      See also |lua-plugin-lazy-loading|.
```

Reply:
>  I would like to ask to rephrase the "intro" or remove it completely due to implied negative connotations

I don't know where you read this in between the lines
I think it's perfectly fine and there is value in describing what the current "de facto standard" is and why that's not good

> Same with using "Cleanly separated ..." (implying that "not separated is not clean"). I'd suggest using either "explicitly separated", "intentionally separated", or just "separated" without any description.

As far as I know as a non-native English speaker "clean separation" doesn't at all imply that otherwise it's not clean (which imo even is the case here but that's not the discussion right now)
as duckduckgo ai chat said:
> The term "clean separation" typically refers to a clear and distinct division or distinction between two things or entities. It does not necessarily imply cleanliness in the literal sense, but rather emphasizes a clear and unambiguous separation.

I found "separated configuration" quite a confusing term though
For me this more sounds like you'd want people to split up the configuration even further

Reply:
> I think it's perfectly fine and there is value in describing what the current "de facto standard" is and why that's not good

Less opinionated -> more better. A suggested rewrite contains both emphasis on configuration-initialization separation and the "de facto standard" which is a perfectly fine alternative if understanding the consequences.

> I found "separated configuration" quite a confusing term though

It is not "<separated configuration> and <initialization>" but more like "configuration and initialization are separate". Both sound OK to me.

Reply:
Given our previous negative interaction, I will give one response and then leave it to everyone else to decide.

I disagree with your suggested change for the following reasons:

It is important to emphasize the consequences of defaulting to forcing users to call a `setup` function in the way I am doing As @max397574 points out, it's a fact that newcomers usually default to this without understanding the consequences.
As mentioned in https://github.com/nvim-neorocks/nvim-best-practices/issues/5#issuecomment-2131367234, plugin authors can choose to diverge if they deem it necessary, which is made clear right at the beginning of this guide:

> it is intended as a set of rules...

I disagree that the way I have phrased it has any negative connotations (this is a nitpick), as I don't make any recommendations not to diverge.

I second @max397574's statement on "clean separation". But I see how this could be misinterpreted as negative for non-native speakers.
How about "strict separation" instead?

Reply:
> It is important to emphasize the consequences of defaulting to forcing users to call a `setup` function

It *is* described in the suggested change: "If you want to force users to explicitly enable plugin functionality, using ...". Without unnecessary "Newcomers often" / "even if they are happy", as built-in docs usually prefer to be more on point.

> How about "strict separation" instead?

To me this is also fine. Using no description is even better.

Reply:
I strongly disagree that your suggestion "is more on point".
The sentence as you suggest would be redundant, as it lacks any reason for it, while losing the *emphasis* on the *very often overlooked* consequences. I have often heard users complain about being forced to call a `setup` function. I have not once had anyone complain about *not being forced* to call one. 

If anything, the phrasing, "Newcomers will often..." makes it clear that this advice is directed at newcomers, knowing that more advanced plugin developers will be able to make their own judgement.

> To me this is also fine. 

I'll adjust it later today.